File: src/main/java/no/fintlabs/model/configuration/ObjectMapping.java
package no.fintlabs.model.configuration;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
public class ObjectMapping {

    public ObjectMapping(
            Map<String, ValueMapping> valueMappingPerKey,
            Map<String, CollectionMapping<ValueMapping>> valueCollectionMappingPerKey,
            Map<String, ObjectMapping> objectMappingPerKey,
            Map<String, CollectionMapping<ObjectMapping>> objectCollectionMappingPerKey
    ) {
        this.valueMappingPerKey = Optional.ofNullable(valueMappingPerKey).orElse(new HashMap<>());
        this.valueCollectionMappingPerKey = Optional.ofNullable(valueCollectionMappingPerKey).orElse(new HashMap<>());
        this.objectMappingPerKey = Optional.ofNullable(objectMappingPerKey).orElse(new HashMap<>());
        this.objectCollectionMappingPerKey = Optional.ofNullable(objectCollectionMappingPerKey).orElse(new HashMap<>());
    }

    private final Map<String, ValueMapping> valueMappingPerKey;
    private final Map<String, CollectionMapping<ValueMapping>> valueCollectionMappingPerKey;
    private final Map<String, ObjectMapping> objectMappingPerKey;
    private final Map<String, CollectionMapping<ObjectMapping>> objectCollectionMappingPerKey;

}

File: src/main/kotlin/no/novari/flyt/mapping/model/configuration/ObjectMapping.kt
package no.novari.flyt.mapping.model.configuration

data class ObjectMapping(
    val valueMappingPerKey: Map<String, ValueMapping> = emptyMap(),
    val valueCollectionMappingPerKey: Map<String, CollectionMapping<ValueMapping>> = emptyMap(),
    val objectMappingPerKey: Map<String, ObjectMapping> = emptyMap(),
    val objectCollectionMappingPerKey: Map<String, CollectionMapping<ObjectMapping>> = emptyMap(),
) {
    class Builder {
        private var valueMappingPerKey: Map<String, ValueMapping>? = null
        private var valueCollectionMappingPerKey: Map<String, CollectionMapping<ValueMapping>>? = null
        private var objectMappingPerKey: Map<String, ObjectMapping>? = null
        private var objectCollectionMappingPerKey: Map<String, CollectionMapping<ObjectMapping>>? = null

        fun valueMappingPerKey(valueMappingPerKey: Map<String, ValueMapping>) =
            apply { this.valueMappingPerKey = valueMappingPerKey }

        fun valueCollectionMappingPerKey(valueCollectionMappingPerKey: Map<String, CollectionMapping<ValueMapping>>) =
            apply { this.valueCollectionMappingPerKey = valueCollectionMappingPerKey }

        fun objectMappingPerKey(objectMappingPerKey: Map<String, ObjectMapping>) =
            apply { this.objectMappingPerKey = objectMappingPerKey }

        fun objectCollectionMappingPerKey(
            objectCollectionMappingPerKey: Map<String, CollectionMapping<ObjectMapping>>,
        ) = apply { this.objectCollectionMappingPerKey = objectCollectionMappingPerKey }

        fun build(): ObjectMapping {
            return ObjectMapping(
                valueMappingPerKey = valueMappingPerKey ?: emptyMap(),
                valueCollectionMappingPerKey = valueCollectionMappingPerKey ?: emptyMap(),
                objectMappingPerKey = objectMappingPerKey ?: emptyMap(),
                objectCollectionMappingPerKey = objectCollectionMappingPerKey ?: emptyMap(),
            )
        }
    }

    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()
    }
}
